feat(theme): support navigating to nested categories in mobile side menu

sethidden opened the pull request:
## Description
Add the ability to enter subcategories from the side menu on mobile
Before this PR you could only access top level categories like Men, Women, Gear, but now you can enter eg. Men>Tops>Jackets

## Related Issue
<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Screenshots (if appropriate):

https://user-images.githubusercontent.com/5359825/157657394-fa370d4a-201a-44c5-a1c1-8cffdb91200f.mov

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

sethidden commented:
Wait for #713 to be merged then fix conflicts

sethidden commented:
1. Move MobileMenuSidebar to the catalog/category module and change its name so its not misleading
2. Remove useSidebar category functions and put it in separate file that's close to MobileCATEGORYMenuSidebar
3. Maybe break down useSidebar and move to dumb helpers module

atagisow-alokai committed:
fix(theme): allow spec.ts files to be used

Before this commit there were 2 issues:
- Types from Cypress and Jest fighting for the `expect()` keyword in .spec.ts files
  but this happened only when running tests with Jest (ts-jest)

- TypeScript didn't know where to find the @types/jest (when using
  tsserver), so I changed the typeRoots to point to the parent pkg

This is only a half-solution as I think if I used some composite configs
it'd be more manageable

fix(theme): squash this

fix(theme): squash this

sethidden commented on packages/theme/modules/catalog/category/components/sidebar/CategorySidebar.vue:
```diff
@@ -83,11 +83,10 @@ export default defineComponent({
 
     onMounted(async () => {
       categoryTree.value = await loadCategoryTree() ?? {};
-      activeCategory.value = findActiveCategory(categoryTree.value, route.value.fullPath);
```
Just a rename since the "findActiveCategory" was misleading - it only finds the label, not the whole object

bartoszherba replied:
Can we return object instead so that function will be more useful? ;)

sethidden replied:
https://github.com/vuestorefront/magento2/pull/709/commits/fa3b0b6cdf90a9d03b72d76adfb14aa465c3390b

sethidden replied:
Done, moved to another file aswell

sethidden commented on packages/theme/modules/catalog/category/components/sidebar/MobileCategorySidebar/MobileCategorySidebar.vue:
```diff
@@ -0,0 +1,129 @@
+<template>
```
Loads of diffs because I just moved the file, but there were plenty of changes to the template and script

sethidden commented on packages/theme/modules/catalog/category/components/sidebar/MobileCategorySidebar/MobileCategorySidebar.vue:
```diff
@@ -0,0 +1,129 @@
+      router.push(path);
+    };
+
+    const activeCategory: CategoryTreeInterface = findDeep(categoryTree.value, (value, key) => key === 'slug' && value === route.value.fullPath.replace('/default/c', ''))?.parent;
```
Note the ugly hack with the `/default/c` replace, but that's the closest I could find to a "unique id" to detect the current category.

sethidden commented on packages/theme/tests/e2e/tsconfig.json:
```diff
@@ -0,0 +1,8 @@
+{
```
This is to fix the long-standing issue with not being able to use spec.ts files. See commit message for more details (the fix(theme) one)

Frodigo replied:
Thanks for that!

sethidden commented:
Hmm the Jira description says `The MobileMenuSidebar component is lazy-loaded`, not sure if that's done

sethidden commented:
`The useCategory and categoryGetters are removed from MobileMenuSidebar componet` this is also not done yet

sethidden commented on packages/theme/components/BottomNavigation.vue:
```diff
@@ -109,12 +113,21 @@ export default defineComponent({
       }
     };
 
+    const loadCategoryMenu = async () => {
```
take notice: moved loading initial data from MobileCategoryMenu itself to BottomNavigation, so MobileCategoryMenu is no longer "self sufficient" (ie. "show and forget")

It was easier to initialize my logic composables this way as the data was already downloaded earlier

sethidden commented on packages/theme/modules/catalog/category/components/sidebar/CategorySidebar.vue:
```diff
@@ -4,17 +4,17 @@
     <SfAccordion
-      :open="activeCategory"
-      :show-chevron="true"
+      :open="topLevelCategoryLabel"
```
Use new findCategoryAncestors function to find the initial top-level category, hopefully this will be more exentsible when we get around to supporting nested categories in the desktop view

Frodigo replied:
I see, thanks!

sethidden commented on packages/theme/modules/catalog/category/components/sidebar/MobileCategorySidebar/logic.ts:
```diff
@@ -0,0 +1,20 @@
+import { CategoryTreeInterface } from '~/modules/catalog/category/types';
+
+export const useMobileCategoryTree = (initialHistory: CategoryTreeInterface[] = []) => {
+  const history = ref<CategoryTreeInterface[]>(initialHistory);
```
There's something really wrong with TypeScript here as it can't understand the <T = CategoryTreeInterface> syntax, so there's lots of duplication of types

sethidden commented on packages/theme/modules/catalog/category/components/sidebar/command/categoryList.gql.ts:
```diff
@@ -1,40 +1,33 @@
 import { gql } from 'graphql-request';
 
+const fragmentCategory = gql`
```
Extract to fragment as not to repeat the same fields all the time

sethidden commented:
need to resolve conflicts, sec

sethidden commented on packages/theme/modules/catalog/category/components/sidebar/CategorySidebar.vue:
```diff
@@ -79,23 +82,31 @@ export default defineComponent({
+    const activeCategoryAncestors = ref(null);
+    const isLoading = ref(true);
 
     onMounted(async () => {
```
Shouldn't this be useAsync?

sethidden replied:
Probably out of scope for this PR, but still

Frodigo replied:
The onMounted is good here, because we won't cache the category tree in Redis/cdn so they are loaded on the client-side.

atagisow-alokai committed:
feat(theme): add nested categories in mobile category menu

refactor(theme): refactor find active category function to return whole object

squash this

feat(theme): pinia working wip

feat(theme): reorganize folders for helpers

feat(theme): reduce duplication in graphql query

feat(theme): adjust desktop and mobile categories

fix(theme): fix invalid import paths
